### PR TITLE
Update signer to support multiple stacks node urls

### PR DESCRIPTION
**Pull request by xoloki**
## Description
Update code and config to support multiple stacks urls.

Closes: #225 

## Changes
Config and stacks api now store multiple endpoints. Stacks api also keeps an index for the current endpoint. This index is incremented on network failure. 

## Testing Information
Existing tests have been updated to check previous and new functionality. 

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: signer/src/stacks/api.rs [hunk 1]
```diff
@@ -250,7 +249,10 @@ impl TryFrom<AccountEntryResponse> for AccountInfo {
 pub struct StacksClient {
     /// The base URL (with the port) that will be used when making requests
     /// for to a Stacks node.
-    pub node_endpoint: url::Url,
+    pub node_endpoints: Vec<url::Url>,
+    /// The current index into the endpoints list
+    /// Increment this on network failure
+    pub endpoint_index: usize,
```
Comment (djordon) [inline signer/src/stacks/api.rs:255]: I think we should make these fields private, since there are certain invariants that the struct needs to uphold. The `node_endpoints` must be non-empty and the `endpoint_index` value is always less than the `node_endpoints`' length.

---

As an aside, you could have a `node_endpoint` field that is still a `url::Url` and a `endpoints` field that is a `std::iter::Cycle<std::vec::IntoIter<url::Url>>`. Getting the next endpoint is then just 
```rust
// You can unwrap because an invariant is the `Vec` that created
// the `IntoIter` was nonempty.
self.node_endpoint = self.endpoints.next().unwrap();
```
Eh, actually, that might be overkill. Getting the endpoints the way that you have here is fine and more obvious if since you don't need to know about `std::iter::Cycle` type.

**Review by djordon — COMMENTED** [review]
This looks okay, but I think we inadvertently introduced some behavior that we don't want. I think I see one main issue, and another "stylistic issue":
1. We switch to another stacks node even when the current stacks node that we were using was perfectly fine. For example, Stacks nodes return an error when they don't have enough data to estimate fees, the node could be perfectly fine. With these changes, we'd change node endpoints in that example, but I think it would be better to change node endpoints only when we suspect that the node is down.
2. We switch to a new node within the endpoint functions when we it feels more natural to switch outside of them. To me, it feels better to have a simple endpoint function that wraps the API call, and another that uses the simple endpoint function and rotates the node endpoint if it gets connectivity error (or a similar error) from the simple function. This other function could also retry the call after switching nodes (but we can save that part for another day). So I wouldn't change the endpoint functions here, probably just the trait functions.

Anyway, let me know if you think that these suggestions make sense.

**Comment by xoloki** [comment]
1. Totally agree. Earlier versions of this PR only rotated urls on connection errors or if the response didn’t complete. But once I realized I could put a closure in map_err, I just did them all 😂
2. Disagree on this. What you’ve described sounds much more complicated, and other than saving a few mutables doesn’t seem to gain us anything.  Logically the API should know the URLs, and it knows when connection errors happen, so it should be the one to rotate the URLs.

File: signer/src/stacks/api.rs [hunk 2]
```diff
@@ -394,9 +426,9 @@ impl StacksClient {
     /// If the given block ID does not exist or is an ID for a non-Nakamoto
     /// block then a Result::Err is returned.
     #[tracing::instrument(skip(self))]
-    async fn get_block(&self, block_id: StacksBlockId) -> Result<NakamotoBlock, Error> {
+    async fn get_block(&mut self, block_id: StacksBlockId) -> Result<NakamotoBlock, Error> {
```
[the remark this answers is not in the document]
Comment (xoloki) [reply signer/src/stacks/api.rs:429]: I really don’t like that idea. I know it has awkward consequences in other parts of the code, but it seems to wildly break encapsulation to do anything other than rotate the URLs inside these functions

File: signer/src/stacks/api.rs [hunk 3]
```diff
@@ -394,9 +426,9 @@ impl StacksClient {
...
         let path = format!("/v3/blocks/{}", block_id.to_hex());
-        let base = self.node_endpoint.clone();
+        let base = self.node_endpoints[self.endpoint_index].clone();
```
[the remark this answers is not in the document]
Comment (xoloki) [reply signer/src/stacks/api.rs:431]: Agreed, will do.

File: signer/src/config.rs [hunk 4]
```diff
@@ -117,7 +120,7 @@ pub struct StacksNodeSettings {
     ///
     /// The endpoint to use when making requests to a stacks node.
     #[serde(deserialize_with = "url_deserializer")]
-    pub endpoint: url::Url,
+    pub endpoints: Vec<url::Url>,
```
[the remark this answers is not in the document]
Comment (xoloki) [reply signer/src/config.rs:123]: Yeah, split should always return at least one string. If it doesn’t convert to a URL then the function fails, otherwise we get a vec with at least one entry.

File: signer/src/stacks/api.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (xoloki) [reply signer/src/stacks/api.rs]: Done

File: signer/src/stacks/api.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (xoloki) [reply signer/src/stacks/api.rs]: Moved endpoint rotation to trait as per our convo

File: signer/src/stacks/api.rs [hunk 5]
```diff
@@ -374,16 +397,25 @@ impl StacksClient {
...
+                self.next_endpoint();
+                Error::StacksNodeResponse(e)
+            })?
             .json()
             .await
-            .map_err(Error::UnexpectedStacksResponse)
+            .map_err(|e| {
+                self.next_endpoint();
+                Error::UnexpectedStacksResponse(e)
+            })
```
[the remark this answers is not in the document]
Comment (xoloki) [reply signer/src/stacks/api.rs]: All trait functions are now rotating endpoint on network errors

File: signer/src/stacks/api.rs [hunk 6]
```diff
@@ -535,25 +586,34 @@ impl StacksClient {
...
     }
 }
 
 impl StacksInteract for StacksClient {
-    async fn get_block(&self, block_id: StacksBlockId) -> Result<NakamotoBlock, Error> {
+    async fn get_block(&mut self, block_id: StacksBlockId) -> Result<NakamotoBlock, Error> {
```
[the remark this answers is not in the document]
Comment (xoloki) [reply signer/src/stacks/api.rs]: Done

File: signer/src/stacks/api.rs (hunk 1 above)
Comment (xoloki) [reply signer/src/stacks/api.rs]: endpoint fields are now private

File: signer/src/config.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (xoloki) [reply signer/src/config.rs:123]: Done

**Comment by xoloki** [comment]
Ok @djordon I've addressed all review comments, and also squashed the bulk of the commits to make the PR less verbose.  lmk what you think.

File: signer/src/stacks/api.rs [hunk 7]
```diff
@@ -547,14 +572,57 @@ impl StacksClient {
 }
 
 impl StacksInteract for StacksClient {
-    async fn get_block(&self, block_id: StacksBlockId) -> Result<NakamotoBlock, Error> {
-        self.get_block(block_id).await
+    async fn get_account(&mut self, address: &StacksAddress) -> Result<AccountInfo, Error> {
+        let result = StacksClient::get_account(self, address).await;
+        match result {
+            Err(Error::StacksNodeRequest(_)) | Err(Error::StacksNodeResponse(_)) => {
```
Comment (djordon) [inline signer/src/stacks/api.rs:578]: It's not clear that you want to match on all `Error::StacksNodeRequest` and all `Error::StacksNodeResponse` variants here. Instead you probably want to check the type of request error in `Error::StacksNodeRequest` and match on whether it is a timeout error, connect error, and maybe some other kind. Matching on all of it is probably too broad, so for now just match on timeout and connect errors.

File: signer/src/stacks/api.rs [hunk 8]
```diff
@@ -565,24 +633,26 @@ impl StacksInteract for StacksClient {
     /// have enough information to provide an estimate, we then get the
     /// current high priority fee for an STX transfer and use that as an
     /// estimate for the transaction fee.
-    async fn estimate_fees<T>(&self, payload: &T, priority: FeePriority) -> Result<u64, Error>
+    async fn estimate_fees<T>(&mut self, payload: &T, priority: FeePriority) -> Result<u64, Error>
```
Comment (djordon) [inline signer/src/stacks/api.rs:636]: I'd just revert the changes here

File: signer/src/block_observer.rs [hunk 9]
```diff
@@ -736,7 +741,18 @@ mod tests {
     }
 
     impl StacksInteract for TestHarness {
-        async fn get_block(&self, block_id: StacksBlockId) -> Result<NakamotoBlock, error::Error> {
+        async fn get_account(&mut self, _address: &StacksAddress) -> Result<AccountInfo, Error> {
+            todo!()
+        }
+
+        async fn submit_tx(&mut self, _tx: &StacksTransaction) -> Result<SubmitTxResponse, Error> {
+            todo!()
```
Comment (djordon) [inline signer/src/block_observer.rs:749]: Whenever we add TODOs like this we've been adding a comment about the ticket that is associated with this TODO. I think we should continue that practice and add such a comment here.

File: signer/src/stacks/api.rs [hunk 10]
```diff
@@ -547,14 +572,57 @@ impl StacksClient {
...
+    }
+    async fn get_tenure_info(&mut self) -> Result<RPCGetTenureInfo, Error> {
+        let result = StacksClient::get_tenure_info(self).await;
+        match result {
+            Err(Error::StacksNodeRequest(_)) | Err(Error::StacksNodeResponse(_)) => {
+                self.next_endpoint();
+            }
+            _ => (),
+        }
+        result
```
Comment (djordon) [inline signer/src/stacks/api.rs:625]: I think it would better to just create a function for this logic. I believe most of these functions can lean on the same logic here.

**Review by djordon — COMMENTED** [review]
Great, I think this feels better. My main comments are around factoring out the identical logic into a function and matching on fewer error variants.

File: signer/src/stacks/api.rs [hunk 11]
```diff
@@ -521,8 +546,8 @@ impl StacksClient {
     /// Uses the GET /v3/tenures/info stacks node endpoint for retrieving
     /// tenure information.
     #[tracing::instrument(skip(self))]
-    pub async fn get_tenure_info(&self) -> Result<RPCGetTenureInfo, Error> {
-        let base = self.node_endpoint.clone();
+    pub async fn get_tenure_info(&mut self) -> Result<RPCGetTenureInfo, Error> {
```
[the remark this answers is not in the document]
Comment (xoloki) [reply signer/src/stacks/api.rs]: Done.

File: signer/src/stacks/api.rs (hunk 7 above)
Comment (xoloki) [reply signer/src/stacks/api.rs]: Done.

File: signer/src/stacks/api.rs (hunk 8 above)
Comment (xoloki) [reply signer/src/stacks/api.rs]: Done.

File: signer/src/stacks/api.rs (hunk 10 above)
Comment (xoloki) [reply signer/src/stacks/api.rs]: Done

File: signer/src/block_observer.rs (hunk 9 above)
Comment (xoloki) [reply signer/src/block_observer.rs:749]: Are we planning on using these functions in the block observer test harness?

Comment (xoloki) [reply signer/src/block_observer.rs]: I tagged them with issue 118, which is to implement StacksInteract for the block observer

Comment (djordon) [reply signer/src/block_observer.rs]: Oh oops, you're right. I don't think we're going to implement these. No need to modify it further, we can remove the TODO and replace it with a comment about how implementing these is unnecessary later.

**Review by djordon — APPROVED** [review]
Looks good ✅
